include/deal.II/distributed/cell_data_transfer.h

```diff
@@ -0,0 +1,346 @@
+                   ExcMessage(
+                     "Values on cells that will be coarsened are not equal!"));
+
+          return value;
+        }
+
+        /**
+         * Evaluate data from @p input_vector on children of @p parent.
+         * Return sum.
+         */
+        static typename VectorType::value_type
+        sum(const typename parallel::distributed::Triangulation<dim, spacedim>::
+              cell_iterator & parent,
+            const VectorType *input_vector)
+        {
+          typename VectorType::value_type sum = 0;
+
+          for (unsigned int child_index = 0; child_index < parent->n_children();
+               ++child_index)
+            sum +=
+              (*input_vector)[parent->child(child_index)->active_cell_index()];
+
+          return sum;
+        }
+
+        /**
+         * Evaluate data from @p input_vector on children of @p parent.
+         * Return mean value.
+         */
+        static typename VectorType::value_type
+        mean(const typename parallel::distributed::
+               Triangulation<dim, spacedim>::cell_iterator &parent,
+             const VectorType *                             input_vector)
+        {
+          return sum(parent, input_vector) / parent->n_children();
+        }
+      };
+
+      /**
+       * Constructor.
+       *
+       * @param[in] triangulation The triangulation on which all operations will
+       *   happen. At the time when this constructor is called, the refinement
+       *   in question has not happened yet.
+       * @param[in] transfer_variable_size_data Specify whether size of data to
+       *   transfer varies from cell to cell.
+       * @param[in] coarsening_strategy Function deciding which data to store on
+       *   a cell, whose children will get coarsened into.
+       */
+      CellDataTransfer(const parallel::distributed::Triangulation<dim, spacedim>
+                         &        triangulation,
+                       const bool transfer_variable_size_data = false,
+                       std::function<typename VectorType::value_type(
+                         const typename parallel::distributed::
+                           Triangulation<dim, spacedim>::cell_iterator &parent,
+                         const VectorType *input_vector)> coarsening_strategy =
+                         &CoarseningStrategies::check_equality);
+
+      /**
+       * Prepare the current object for coarsening and refinement. It registers
+       * the data transfer of @p in on the underlying triangulation. @p in includes
+       * data to be interpolated onto the new (refined and/or coarsened) grid.
+       */
+      void
+      prepare_for_coarsening_and_refinement(const VectorType &in);
+
+      /**
+       * Same as the function above, only for a list of vectors.
+       */
+      void
+      prepare_for_coarsening_and_refinement(
+        const std::vector<const VectorType *> &all_in);
```

Comment:
Can we find an alternative solution for this interface?

I am not a huge fan of `std::vector<const VectorType *> &` I have to say. Further, `prepare_for_coarsening_and_refinement` to register a (read-only) vector and `unpack` to read out the transferred values is not very intuitive.

What about the following high-level interface:
 - A single `register(VectorType &vec)`, and corresponding `deregister(VectorType &vec)` function that registers vectors that shall be maintained by the CellDataTransfer object.
 - When you register a number of vectors `vec1`, `vec2`, etc. the class automatically
   - makes a copy of the data before coarsening/refinement
   - resizes the referenced vectors after coarsening/refinement and writes out the new values immediately to `vec1`, `vec2`, and removes the temporary storage?

This would significantly simplify the interface :-)

Reply:
Later, if you want to register a container storing vectors you can simply do:
```
for(auto & it : container_with_vectors)
  cell_data_transfer.register(it);
```
(or use `std::for_each`). There is no need to add inflexible wrappers such as `std::vector<VectorType *>` in the calling interface.

Reply:
@tamiko -- that's a design decision copied from the `SolutionTransfer` class. We made this mistake 20 years ago, and I'm in favor of using the same interface in all of these classes, rather than breaking it for one class but leaving the others as is.

Reply:
Well, then let's go that route. (And think about a better design later.)